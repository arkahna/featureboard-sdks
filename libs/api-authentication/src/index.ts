export {
    readCurrentOrganization,
    writeCurrentOrganization,
} from './lib/current-organization'
export { getValidToken } from './lib/get-valid-token'
export {
    TOKEN_URL,
    performTokenRefresh,
    readToken,
    writeToken,
} from './lib/token'
export type { TokenData } from './lib/token'
