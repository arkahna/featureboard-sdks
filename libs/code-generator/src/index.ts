export { getUserOrganizations } from './lib/api/get-organizations'
export { getProjectFeatures } from './lib/api/get-project-features'
export type { FeatureBoardFeature } from './lib/api/get-project-features'
export { getProjects } from './lib/api/get-projects'
export { codeGenerator } from './lib/code-generator'
export type { CodeGeneratorOptions, Template } from './lib/code-generator'
export { queryFeatureBoard } from './lib/queryFeatureBoard'
export type { FeatureBoardAuth } from './lib/queryFeatureBoard'
export { FsTree, flushChanges, printChanges } from './lib/tree/tree'
export type { Tree } from './lib/tree/tree'
