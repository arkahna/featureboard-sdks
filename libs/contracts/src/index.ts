export * from './lib/error-types'
export * from './lib/messages'
export * from './lib/notification-types'
