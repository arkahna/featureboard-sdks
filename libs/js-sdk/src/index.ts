export type { BrowserClient } from './client-connection'
export { createBrowserClient } from './create-browser-client'
export { createManualClient } from './create-manual-client'
export type { FeatureBoardApiConfig } from './featureboard-api-config'
export type { FeatureBoardClient } from './features-client'

// TODO We should make these 'internal' and not export them
// Need to figure out how to do that with the current build setup
export { createEnsureSingle } from './ensure-single'
export { featureBoardHostedService } from './featureboard-service-urls'
export { retry } from './utils/retry'

export interface Features {}
