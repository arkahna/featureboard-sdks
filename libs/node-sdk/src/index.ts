export type {
    FeatureBoardApiConfig,
    FeatureBoardClient,
    Features,
} from '@featureboard/js-sdk'
export type { ExternalStateStore } from './external-state-store'
export { createManualServerClient } from './manual-server-client'
export { createServerClient } from './server-client'
export type { CreateServerClientOptions } from './server-client'
export type { ServerClient } from './server-connection'
