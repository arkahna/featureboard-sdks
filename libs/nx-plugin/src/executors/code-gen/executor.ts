import {
    getValidToken,
    readCurrentOrganization,
} from '@featureboard/api-authentication'
import {
    codeGenerator,
    type FeatureBoardAuth,
} from '@featureboard/code-generator'
import { joinPathFragments, type ExecutorContext, type Tree } from '@nx/devkit'
import { FsTree, flushChanges, printChanges } from 'nx/src/generators/tree'
import { API_ENDPOINT, CLIENT_ID } from '../../shared/config'
import type { CodeGenExecutorSchema } from './schema'

export async function codeGenExecutor(
    options: CodeGenExecutorSchema,
    context: ExecutorContext,
    tree?: Tree,
) {
    console.log('codeGenExecutor', options)
    if (!context.projectName) {
        throw new Error('No projectName: name not found in in project.json')
    }

    const projectRoot =
        context.projectsConfigurations?.projects[context.projectName]?.root

    if (!projectRoot) {
        console.error(`Error: Cannot find root for ${context.projectName}.`)
        return {
            success: false,
        }
    }

    let auth: FeatureBoardAuth
    if (options.featureBoardApiKey) {
        auth = {
            featureBoardApiKey: options.featureBoardApiKey,
        }
    } else {
        const featureBoardBearerToken = await getValidToken(CLIENT_ID)
        const organizationId = await readCurrentOrganization(context.isVerbose)
        if (!featureBoardBearerToken || !organizationId) {
            return
        }

        auth = {
            featureBoardBearerToken,
            organizationId,
        }
    }

    tree = tree || new FsTree(projectRoot, context.isVerbose)

    await codeGenerator({
        tree: tree,
        relativeFilePath: joinPathFragments(projectRoot, options.subFolder),
        interactive: false,
        auth,
        apiEndpoint: API_ENDPOINT,
        ...options,
    })
    saveChanges(tree, options.dryRun, context.isVerbose)

    return {
        success: true,
    }
}

function saveChanges(tree: Tree, dryRun: boolean, verbose: boolean) {
    const changes = tree.listChanges()
    if (verbose) {
        printChanges(changes)
        if (changes.length === 0) {
            console.log(`\nFiles are the same no changes were made.`)
            return
        }
    }

    if (!dryRun && changes.length !== 0) {
        flushChanges(tree.root, changes)
        return
    }
    console.log(`\nNOTE: The "dryRun" flag means no changes were made.`)
}
export default codeGenExecutor
