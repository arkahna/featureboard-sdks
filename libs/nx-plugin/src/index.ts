export * from './executors/code-gen/executor'
export * from './generators/code-gen/generator'
