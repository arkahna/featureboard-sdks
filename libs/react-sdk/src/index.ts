export * from './featureboard-provider'
export * from './use-feature'
export * from './use-client'
