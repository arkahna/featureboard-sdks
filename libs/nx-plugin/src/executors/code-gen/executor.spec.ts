import type { ExecutorContext, Tree } from '@nx/devkit'
import { addProjectConfiguration, joinPathFragments } from '@nx/devkit'
import * as fs from 'fs/promises'
import { HttpResponse, http } from 'msw'
import type { SetupServer } from 'msw/node'
import { setupServer } from 'msw/node'
import { createTreeWithEmptyWorkspace } from 'nx/src/devkit-testing-exports'
import * as path from 'path'
import { codeGenExecutor } from './executor'
import type { CodeGenExecutorSchema } from './schema'

describe('code-gen executor', () => {
    let context: ExecutorContext
    let tree: Tree
    let options: CodeGenExecutorSchema
    let server: SetupServer
    const root = '/virtual'
    const projectName = 'my-app'
    const projectRoot = `apps/${projectName}`

    beforeAll(() => {
        server = setupServer(
            http.get('https://api.featureboard.app/projects', async () => {
                const file = await fs.readFile(
                    path.join(__dirname, '../../../test-data/projects.json'),
                    {
                        encoding: 'utf8',
                    },
                )
                return HttpResponse.json(JSON.parse(file))
            }),
            http.get(
                'https://api.featureboard.app/projects/ivu5ntxbxeeorm248jl25bs6/features',
                async () => {
                    const file = await fs.readFile(
                        path.join(
                            __dirname,
                            '../../../test-data/features-ivu5ntxbxeeorm248jl25bs6.json',
                        ),
                        {
                            encoding: 'utf8',
                        },
                    )
                    return HttpResponse.json(JSON.parse(file))
                },
            ),
        )
        server.listen({ onUnhandledRequest: 'error' })
    })

    beforeEach(async () => {
        context = {
            root: root,
            cwd: root,
            projectName: projectName,
            targetName: 'lint',
            projectsConfigurations: {
                version: 2,
                projects: {
                    [projectName]: {
                        root: projectRoot,
                        sourceRoot: projectRoot,
                        targets: {
                            lint: {
                                executor: '@nx-dotnet/core:format',
                            },
                        },
                    },
                },
            },
            isVerbose: false,
        }

        options = {
            subFolder: './src/features',
            template: 'dotnet-api',
            featureBoardProductName: 'SaaSy Icons',
            featureBoardApiKey: 'This is totally a key',
            dryRun: true,
        }

        tree = createTreeWithEmptyWorkspace({ layout: 'apps-libs' })
        addProjectConfiguration(tree, projectName, {
            root: projectRoot,
            projectType: 'application',
            targets: { 'my-target': { executor: 'nx:noop' } },
        })
        tree.write(
            joinPathFragments(projectRoot, 'Project.FeatureBoard.MyApp.csproj'),
            '',
        )
    })

    describe('template: dotnet-api', () => {
        beforeEach(async () => {
            options.template = 'dotnet-api'
        })

        it('should run successfully', async () => {
            const output = await codeGenExecutor(options, context, tree)

            expect(output?.success).toBe(true)
        })

        it('should produce the expected features files', async () => {
            await codeGenExecutor(options, context, tree)
            const files = tree
                .listChanges()
                .filter(
                    (x) =>
                        x.path.match(/^apps\/my-app\/.*$/) &&
                        !x.path.endsWith('.csproj') &&
                        !x.path.endsWith('/project.json'),
                )

            expect(files).toHaveLength(1)
            files.forEach((x) => {
                expect(x.content?.toString('utf-8')).toMatchSnapshot(x.path)
            })
        })
    })

    describe('template: typescript', () => {
        beforeEach(async () => {
            options.template = 'typescript'
        })

        it('should run successfully', async () => {
            const output = await codeGenExecutor(options, context, tree)

            expect(output?.success).toBe(true)
        })

        it('should produce the expected features files', async () => {
            await codeGenExecutor(options, context, tree)
            const files = tree
                .listChanges()
                .filter(
                    (x) =>
                        x.path.match(/^apps\/my-app\/.*$/) &&
                        !x.path.endsWith('.csproj') &&
                        !x.path.endsWith('/project.json'),
                )

            expect(files).toHaveLength(1)
            files.forEach((x) => {
                expect(x.content?.toString('utf-8')).toMatchSnapshot(x.path)
            })
        })
    })

    afterEach(() => {
        server.resetHandlers()
    })

    afterAll(() => {
        server.close()
    })
})
