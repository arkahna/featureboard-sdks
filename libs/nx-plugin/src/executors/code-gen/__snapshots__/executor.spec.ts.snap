// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`code-Executor > template: dotnet-api > should produce the expected features files > apps/my-app/Features.cs 1`] = `
"using FeatureBoard.DotnetSdk.Attributes;
using FeatureBoard.DotnetSdk.Models;
using System.Runtime.Serialization;

namespace Project.FeatureBoard.MyApp;

// Features
public class Features : IFeatures
{
  [FeatureKeyName(\\"requires-attribution\\")]
  public bool RequiresAttribution { get; set; }
  [FeatureKeyName(\\"allow-edits\\")]
  public bool AllowEdits { get; set; }
  [FeatureKeyName(\\"number-custom-collections\\")]
  public decimal NumberCustomCollections { get; set; }
  [FeatureKeyName(\\"team-management\\")]
  public bool TeamManagement { get; set; }
  [FeatureKeyName(\\"custom-collections\\")]
  public CustomCollections CustomCollections { get; set; }
  [FeatureKeyName(\\"view-subscribers\\")]
  public bool ViewSubscribers { get; set; }
}


// Options Enums

//Custom collections
public enum CustomCollections
{
  [EnumMember(Value = \\"Disabled\\")]
  Disabled,
  [EnumMember(Value = \\"Enabled\\")]
  Enabled,
  [EnumMember(Value = \\"Shared\\")]
  Shared,
}

// Type Enums
public enum BooleanFeature
{
  RequiresAttribution,
  AllowEdits,
  TeamManagement,
  ViewSubscribers,
}
public enum NumberFeature
{
  NumberCustomCollections,
}
public enum StringFeature
{
}

// Feature Gate attribute
[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = true)]
public class FeatureFilterAttribute : FeatureFilterAttributeBase
{
  protected override string Feature { get; }
  protected override object DefaultValue { get; }
  protected override object? IsEqualTo { get; }
  private readonly Dictionary<string, string> _featureKeyLookup = new()
  {
    { \\"RequiresAttribution\\", \\"requires-attribution\\" },
    { \\"AllowEdits\\", \\"allow-edits\\" },
    { \\"NumberCustomCollections\\", \\"number-custom-collections\\" },
    { \\"TeamManagement\\", \\"team-management\\" },
    { \\"ViewSubscribers\\", \\"view-subscribers\\" },
  };

  public FeatureFilterAttribute(BooleanFeature feature, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(BooleanFeature feature, bool isEqualTo, bool defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(StringFeature feature, string isEqualTo, string defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(NumberFeature feature, decimal isEqualTo, decimal defaultValue)
  {
    Feature = _featureKeyLookup.GetValueOrDefault(feature.ToString()) ?? feature.ToString();
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

  public FeatureFilterAttribute(CustomCollections isEqualTo, CustomCollections defaultValue)
  {
    Feature = \\"custom-collections\\";
    IsEqualTo = isEqualTo;
    DefaultValue = defaultValue;
  }

}
"
`;

exports[`code-Executor > template: dotnet-api > should produce the expected features files > apps/my-app/Project.FeatureBoard.MyApp.csproj 1`] = `""`;

exports[`code-Executor > template: dotnet-api > should produce the expected features files > apps/my-app/project.json 1`] = `
"{
  \\"name\\": \\"my-app\\",
  \\"$schema\\": \\"../../node_modules/nx/schemas/project-schema.json\\",
  \\"projectType\\": \\"application\\",
  \\"targets\\": {
    \\"my-target\\": {
      \\"executor\\": \\"nx:noop\\"
    }
  }
}
"
`;

exports[`code-Executor > template: typescript > should produce the expected features files > apps/my-app/Project.FeatureBoard.MyApp.csproj 1`] = `""`;

exports[`code-Executor > template: typescript > should produce the expected features files > apps/my-app/featureboard-features.ts 1`] = `
"export interface SaaSyIconsFeatures {

    // Requires attribution
    // 
    'requires-attribution': boolean

    // Allow edits
    // Allows the user to edit colour, backgrounds
    'allow-edits': boolean

    // Number custom collections
    // 
    'number-custom-collections': number

    // Team management
    // 
    'team-management': boolean

    // Custom collections
    // 
    'custom-collections': 'Disabled' | 'Enabled' | 'Shared'

    // View subscribers
    // 
    'view-subscribers': boolean
}
export type SaaSyIconsFeature = keyof SaaSyIconsFeatures"
`;

exports[`code-Executor > template: typescript > should produce the expected features files > apps/my-app/project.json 1`] = `
"{
  \\"name\\": \\"my-app\\",
  \\"$schema\\": \\"../../node_modules/nx/schemas/project-schema.json\\",
  \\"projectType\\": \\"application\\",
  \\"targets\\": {
    \\"my-target\\": {
      \\"executor\\": \\"nx:noop\\"
    }
  }
}
"
`;
